cmd/app/start.go

```diff
@@ -43,12 +57,59 @@ var RootCmd = &cobra.Command{
 	Use:   "cert-manager-csi",
 	Short: "Container Storage Interface driver to issue certificates from Cert-Manager",
 	RunE: func(cmd *cobra.Command, args []string) error {
-		d, err := driver.New(opts.DriverName, opts.NodeID, opts.Endpoint, opts.DataRoot, opts.TmpfsSize)
+
+		log := klogr.New()
+
+		restConfig, err := rest.InClusterConfig()
+		if err != nil {
+			panic("cannot load in-cluster config")
+		}
+
+		store, err := storage.NewFilesystem(log, opts.DataRoot)
 		if err != nil {
-			return err
+			panic("failed to setup filesystem: " + err.Error())
+		}
+
+		keyGenerator := keygen.Generator{Store: store}
+		writer := filestore.Writer{Store: store}
+
+		d, err := driver.New(opts.Endpoint, log, driver.Options{
+			DriverName:    opts.DriverName,
+			DriverVersion: "v0.0.1",
+			NodeID:        opts.NodeID,
+			Store:         store,
+			Manager: manager.NewManagerOrDie(manager.Options{
+				Client:             cmclient.NewForConfigOrDie(restConfig),
+				MetadataReader:     store,
+				Clock:              clock.RealClock{},
+				Log:                log,
+				NodeID:             opts.NodeID,
+				GeneratePrivateKey: keyGenerator.KeyForMetadata,
+				GenerateRequest:    requestgen.RequestForMetadata,
+				SignRequest:        signRequest,
+				WriteKeypair:       writer.WriteKeypair,
+			}),
+		})
+		if err != nil {
+			return fmt.Errorf("failed to setup driver: " + err.Error())
+		}
+
+		if err := d.Run(); err != nil {
+			return fmt.Errorf("failed running driver: " + err.Error())
 		}
 
-		d.Run()
 		return nil
 	},
 }
+
+func signRequest(_ metadata.Metadata, key crypto.PrivateKey, request *x509.CertificateRequest) ([]byte, error) {
+	csrDer, err := x509.CreateCertificateRequest(rand.Reader, request, key)
+	if err != nil {
+		return nil, err
+	}
+
+	return pem.EncodeToMemory(&pem.Block{
+		Type:  "CERTIFICATE REQUEST",
+		Bytes: csrDer,
+	}), nil
+}
```

Comment:
suggestion: the comment for SignRequestFunc suggests the returned data should be DER, but here we return PEM. Which one is correct?

(I've _long_ wanted to use explicit types for PEM data and DER data everywhere to make this clearer:

```go
// this is the bare minimum, obviously there are more things that can be done here
type PEMData string
type DERData []byte
```

so that stuff like this is caught at compile time and doesn't have to be left to reviews. It might be a bit late for cert-manager but it could be worth adding down the road here.)

Reply:
Should be PEM as per: https://github.com/cert-manager/csi-lib/blob/fe9b5eff41b7a867e93102b4e7cf7847b40aa23c/manager/interfaces.go#L72-L78 (to allow for arbitrary additional bytes in between PEM blocks if a CSR requestor were to choose to include them)

Reply:
+1 to specific types, however `PEMData` would also need to be a `[]byte` :)

Reply:
This should be PEM format. It used to be DER in an earlier version of csi-lib.

Reply:
>  +1 to specific types, however PEMData would also need to be a []byte :)

Why? Encoded PEM is _explicitly_ a string, that's the whole reason it was created; it's from privacy enhanced mail, and the whole point is that it can allow binary data be sent in an email. This is something the go standard lib gets wrong too, IMO. A valid PEM block is specifically a subset of valid strings which is in turn a subset of valid byte slices.

Is there some technical reason that specifically it has to be a []byte in go?

Reply:
A PEM file can contain multiple concatenated PEM blocks each with a (non-textual) comment between them - if we restrict this to `string`, it won't be possible for anyone to use non-text data between PEM blocks

Reply:
Ah, we're talking about different things then; I'm talking about parsed PEM data, you're talking about files read from disk (which you're correct to point out has to be `[]byte`). 

An ideal function signature with `type PEMData string` to me looks like this:

```go
func ProcessPEMFile(data []byte) ([]PEMData, error)
```

Reply:
In this instance though the CSI driver implementor may need to append additional blocks of data (as this is a function that takes a `*x509.CertificateRequest` and returns the data that should be put into the cert-manager CertificateRequest object), so we need it to be a `[]byte` to permit those implementors that need to make use of additional data between/before/after PEM blocks :)

Reply:
this is my own fault for suggesting something in an aside to the original comment. I wasn't suggesting that this _specific_ function should be changed to anything in particular, I was just saying in general that